archaius2-core/src/test/java/com/netflix/archaius/ProxyFactoryTest.java

```diff
@@ -593,31 +593,18 @@ interface CustomObject {
     @Test
     public void testNestedInterfaceWithCustomDecoder() {
         TypeConverter<ConfigWithNestedInterface.CustomObject> customObjectTypeConverter = value -> value::toUpperCase;
-
-        final class CustomDecoder implements Decoder, TypeConverter.Registry {
-            @Override
-            public <T> T decode(Class<T> type, String encoded) {
-                if (type.equals(ConfigWithNestedInterface.CustomObject.class)) {
-                    @SuppressWarnings("unchecked")
-                    T converted = (T) customObjectTypeConverter.convert(encoded);
-                    return converted;
-                }
-                return DefaultDecoder.INSTANCE.decode(type, encoded);
-            }
-
-            @Override
-            public Optional<TypeConverter<?>> get(Type type) {
-                if (type.equals(ConfigWithNestedInterface.CustomObject.class)) {
-                    return Optional.of(customObjectTypeConverter);
-                }
-                return DefaultDecoder.INSTANCE.get(type);
+        TypeConverter.Factory customTypeConverterFactory = (type, registry) -> {
+            if (type.equals(ConfigWithNestedInterface.CustomObject.class)) {
+                return Optional.of(customObjectTypeConverter);
             }
-        }
+            return Optional.empty();
+        };
+        DefaultDecoder customDecoder = DefaultDecoder.newCustomDecoder(Collections.singletonList(customTypeConverterFactory));
```

Comment:
Naming here is ... unfortunate. Do you feel like it would be worthwhile to rename the classes a bit? Maybe move the existing code to a `CustomizableDecoder` that has your static factory and making `DefaultDecoder` be a sublcass of it that calls the constructor with an empty factory list?